pkg/pillar/cmd/zedagent/handlentp.go

```diff
@@ -0,0 +1,336 @@
+// Copyright (c) 2024 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package zedagent
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+
+	chrony "github.com/facebook/time/ntp/chrony"
+	"github.com/lf-edge/eve-api/go/info"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"google.golang.org/protobuf/proto"
+	timestamppb "google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const (
+	// Path to unix chrony socket
+	unixChronydPath = "/run/chrony/chronyd.sock"
+)
+
+var prevInfoNTPSourcesMap map[destinationBitset]*info.ZInfoNTPSources
+
+func getForceSendInterval(ctx *zedagentContext) time.Duration {
+	interval := ctx.globalConfig.GlobalValueInt(types.NTPSourcesInterval)
+	return time.Duration(interval) * time.Second
+}
+
+// Run a periodic post of the NTP sources information.
+func ntpSourcesTimerTask(ctx *zedagentContext, handleChannel chan interface{},
+	triggerNTPSourcesInfo chan destinationBitset) {
+
+	prevInfoNTPSourcesMap = make(map[destinationBitset]*info.ZInfoNTPSources)
+
+	// Ticker for periodic publishing NTP sources to the controller.
+	// The period is rather small, because NTP sources are published
+	// if something was really changed. See the @publishNTPSourcesToDest
+	ticker := time.NewTicker(10 * time.Second)
+
+	// Return handles to the caller.
+	handleChannel <- ticker
+
+	wdName := agentName + "-ntp"
+
+	// Run a periodic timer so we always update StillRunning
+	stillRunning := time.NewTicker(25 * time.Second)
+	ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	ctx.ps.RegisterFileWatchdog(wdName)
+
+	ts := time.Now()
+	for {
+		select {
+		case <-ticker.C:
+			dest := ControllerDest
+			if time.Now().Sub(ts) > getForceSendInterval(ctx) {
+				// Periodically do a force send to be sure
+				// controller gets an NTP update in any case
+				dest |= ForceSend
+				ts = time.Now()
+			}
+			publishNTPSources(ctx, wdName, dest)
+		case dest := <-triggerNTPSourcesInfo:
+			publishNTPSources(ctx, wdName, dest)
+		case <-stillRunning.C:
+		}
+		ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	}
+}
+
+func publishNTPSources(ctx *zedagentContext, wdName string,
+	dest destinationBitset) {
+	info := getNTPSourcesInfo(ctx)
+	if info == nil {
+		// Not available.
+		return
+	}
+	start := time.Now()
+	publishNTPSourcesToDest(ctx, info, dest)
+	ctx.ps.CheckMaxTimeTopic(wdName, "publishNTPSources", start,
+		warningTime, errorTime)
+}
+
+func ntpSourcesChanged(info *info.ZInfoNTPSources,
+	dest destinationBitset) bool {
+
+	prev, cached := prevInfoNTPSourcesMap[dest]
+	prevInfoNTPSourcesMap[dest] = info
+	if !cached {
+		// First time
+		return true
+	}
+	if len(prev.Sources) != len(info.Sources) {
+		// Set of sources has been changed
+		return true
+	}
+	for i, source := range info.Sources {
+		prevSource := prev.Sources[i]
```

Comment:
Does golang ensure that added sources are inserted in a map such that later additions appear after older additions as seen by the range operator?
Thus if you add the sources in order A, B, C, then will the range return them in order A, B, C?

The reason I'm asking is that golang intentionally makes maps behave in random ways.
(Or is info.Sources an array and not a map?)

Reply:
It is an array.